### PR TITLE
Incremental builds

### PR DESCRIPTION
**Pull request by ddollar**
* scan local directory and take sha256 of every file
* ask server which hashes it doesn't know about
* upload the file content for the missing hashes
* capture current directory as a manifest of hashes to (filename, mode, mtime)
* upload manifest to server
* recreate source directory on the server side using manifest and hashed files
* build using that

## Release Playbook
- [x] Rebase against master
- [x] Pass checks
- [x] Release branch (20160226025613-incremental-builds)
- [x] Pass CI (https://circleci.com/gh/convox/rack/589)
- [x] Code review
- [ ] Merge into master
- [ ] Release master ()
- [ ] Pass CI ()
- [ ] Update dev and testing racks
- [ ] If rack or app formation changes, test upgrade → re-deploy existing app → downgrade cycle
- [ ] Publish release
- [ ] Release CLI

### CODE REVIEW

**Comment by ddollar** [comment]
`20160225030413-incremental-builds`

File: api/controllers/builds.go [hunk 1]
```diff
@@ -134,7 +135,30 @@ func BuildCreate(rw http.ResponseWriter, r *http.Request) *httperr.Error {
...
+		fmt.Printf("build %+v\n", build)
+
+		if err != nil {
+			return httperr.Server(err)
+		} else {
+			return RenderJson(rw, build)
+		}
+	}
+
+	return httperr.Errorf(403, "no source or repo")
```
Comment (nzoschke) [inline api/controllers/builds.go]: Did you intend to remove the tracking here?

Comment (ddollar) [inline api/controllers/builds.go]: No. This is all new code so it just never had it.

File: api/models/build.go [hunk 2]
```diff
@@ -284,6 +285,58 @@ func (b *Build) ExecuteRemote(repo string, cache bool, config string, ch chan er
...
+	args = append(args, b.App, "-")
+
+	err = b.execute(args, bytes.NewReader(tgz), ch)
+
+	if err != nil {
+		b.buildError(err, ch)
+		return
+	}
+
+	NotifySuccess("build:create", map[string]string{"id": b.Id, "app": b.App})
```
Comment (nzoschke) [inline api/models/build.go]: I'd like to TrackSuccess here too to help measure QoS for this path

Comment (ddollar) [inline api/models/build.go]: Added

File: api/controllers/builds.go (hunk 1 above)
Comment (ddollar) [inline api/controllers/builds.go]: Added

**Comment by nzoschke** [comment]
The outline sounds solid. This is a lot like [Anvil](https://github.com/ddollar/anvil), right? That worked very well.  

I'm having a hard time grokking all the code on my first pass. Seems like getting the details perfect is really important.

Can you add some tests that demonstrate and validate how the protocol and API works?

It doesn't block this strategy, but my idea for efficient data transfer is to use the standard git protocol. We could provision a CodeCommit repo for every app, and use an single `httputil.NewSingleHostReverseProxy` handler to proxy with the right auth.

* Extremely efficient transport via the Smart HTTP Git protocol
* Application data encrypted at rest
* Standard .gitignore

**Comment by xtfxme** [comment]
Is there some kind of sync library in Go? eg. an `rsync`-like thing (or git as previously noted)? It seems unfortunate to not lean on some existing, well-established file-transfer proto.

**Comment by nzoschke** [comment]
This is always my goal too:

> lean on some existing, well-established $X

This protocol and implementation is simple and does have prior art in Anvil so theres no question it works.

But if there is a well-established client and server we could use I'll always push for that over writing a custom client and API.

**Comment by ddollar** [comment]
@nzoschke just for clarity are you asking to block the PR on those tests?

**Comment by nzoschke** [comment]
If this is an experiment it's ok without tests

**Comment by nzoschke** [comment]
is there a way to fall back to the tarball path?

**Comment by ddollar** [comment]
the api path still exists so old clients will continue to use it

there's no cli flag or anything to toggle

**Comment by nzoschke** [comment]
I think that deserves more consideration... 

There's an API compatibility issue there (and the version check stuff is removed completely now?) And if a bug shows up in this path, it'd be nice to have the option of going back to the old path

**Comment by ddollar** [comment]
Yep. All of the server side bits are still there. You could grab an older CLI release or we could quick-release a reversion to the CLI if that became necessary.

The version check stuff deserves more thought. We need something much lighter-weight.

**Comment by ddollar** [comment]
Actually once this has a release number I can just make it use the old client code path automatically if the rack version is too old.

**Comment by ddollar** [comment]
Added a `--classic` option for old-style builds. Will remove in a few weeks.

**Pull request merged**